### ISSUE TITLE
[BUG] xgrammar does not follow the constraint

### ISSUE THREAD

**Issue by merrymercy** [sgl-project/sglang#2017]
xgrammar does not follow the integer constraint and generate a floating number for an integer filed.

## Schema
https://github.com/sgl-project/sglang/blob/1f4514601e4c6595bb7b79dd24347ad01fa1d119/test/srt/test_json_constrained.py#L30

## Error
```
======================================================================
FAIL: test_json_openai (__main__.TestJSONConstrained)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/sglang/test/srt/test_json_constrained.py", line 109, in test_json_openai
    assert isinstance(js_obj["population"], int), f"{js_obj=}"
AssertionError: js_obj={'name': 'Paris', 'population': 2.0}
```

## Reproduce:

branch: https://github.com/sgl-project/sglang/tree/xgrammar-fail

```
python3 test/srt/test_json_constrained.py TestJSONConstrained.test_json_openai
```

**Comment by merrymercy** [comment]
cc @Ubospica @DarkSharpness

**Comment by roG0d** [comment]
Hi, I'm researching inference-serving frameworks that implement Grammar-Constrained Decoding (GCD) for Context Free Grammar (CFG) and you guys are the first to include the recent xgrammar. I'm trying to reproduce this error with the specified branch but I get the following new one:
 ``` 
 File "/home/rog0d/miniconda3/envs/sglang/lib/python3.10/site-packages/sglang/srt/constrained/bnf_cache.py", line 44, in __init__
 self.grammar_cache = GrammarMatcherInitContextCache(
TypeError: Dummy() takes no arguments
```
When deploying the container provided [here](https://sgl-project.github.io/start/install.html#method-3-using-docker) with xgrammar as the backend, 
```
docker run --gpus all \
    --shm-size 32g \
    -p 30000:30000 \
    -v ~/.cache/huggingface:/root/.cache/huggingface \
    --env "HF_TOKEN=<hf_token>" \
    --ipc=host \
    lmsysorg/sglang:latest \
    python3 -m sglang.launch_server --model-path meta-llama/Llama-3.2-3B-Instruct --host 0.0.0.0 --port 30000 --grammar-backend xgrammar
```
I get the same Dummy() error.  Is there something different I'm missing? I've traced it back to this [PR](https://github.com/sgl-project/sglang/pull/1752/files#diff-b706610abd4146c4bc7b2c301532a335d13aa98d162af3429c07fda57f4bc538R44).
I would like to test SGLang with CFG (apart from JSON and Regex) using outlines and xgrammar, if I manage to test it I'll gladly contribute it here. 

One possible solution could be to reproduce vLLM OpenAI compatible server and its workflow as they have a [guided_grammar](https://github.com/vllm-project/vllm/blob/ac49b59d8b01ffb9979e18e67b252d45410bc1e6/vllm/entrypoints/openai/protocol.py#L275) parameter.

**Comment by merrymercy** [comment]
@roG0d You need to compile and install xgrammar: https://github.com/mlc-ai/xgrammar

```
pip install pybind11 ninja

mkdir build
cd build
cmake .. -G Ninja
ninja
cd ../python
pip install -e .
```

**Comment by Ubospica** [comment]
@roG0d Yes it is not uploaded to pypi yet. We will further integrate it soon and avoid this compile step.

**Comment by alanxmay** [comment]
xgrammar: https://pypi.org/project/xgrammar/

**Comment by Ubospica** [comment]
That issue is fixed in XGrammar v0.1.5. Now it will only generate integers for integer constraints.
